### PR TITLE
chore: limit the amount of unannounced events we announce

### PR DESCRIPTION
**Pull request by gastonfournier**
## About the changes
When the events table is large we might be doing a full table scan searching for unannounced events. We spotted it due to a performance alert and confirmed in AWS performance insights
image

The proposal is to limit this operation to 500 events (rule of thumb) per round https://github.com/Unleash/unleash/blob/f82ae354ebe6d3b0a3ddf15a051d6c8ed995c10e/src/lib/services/index.ts#L141-L147 and also ignore the events older than a day (because it seems reasonable)

## Discussion points
**Idea**: split the `events` table into `recent_events` and `historical_events`. Recent can be anything from a day/week/month. This would help with recurrent queries that rely on recent data from the event's table such as optimal 304 calculation or event this scheduled task that sends unannounced events.

### CODE REVIEW

**Review by thomasheartman — APPROVED** [review]
This sounds like a sensible change to me! It also only deals with events in the last hour, right? 

As for the discussion point: yeah, that might be a good idea, actually. I'm in favor if it'll have net positive effects.

File: src/lib/db/event-store.ts [hunk 1]
```diff
@@ -412,7 +412,9 @@ class EventStore implements IEventStore {
             .update({ announced: true })
             .where('announced', false)
             .whereNotNull('announced')
-            .returning(EVENT_COLUMNS);
+            .where('created_at', '>', new Date(Date.now() - 1000 * 60 * 60))
```
Comment (gastonfournier) [inline src/lib/db/event-store.ts:415]:
```suggestion
            .where('created_at', '>', new Date(Date.now() - 24 * 1000 * 60 * 60))
```

**Comment by gastonfournier** [comment]
> This sounds like a sensible change to me! It also only deals with events in the last hour, right?
> 
> As for the discussion point: yeah, that might be a good idea, actually. I'm in favor if it'll have net positive effects.

Actually, I wanted to limit this to last 24hs :P good spot: https://github.com/Unleash/unleash/pull/4845/files#r1338329830

File: src/lib/db/event-store.ts (hunk 1 above)
Comment (kwasniew) [reply src/lib/db/event-store.ts:415]: I'd use: `import { addDays, subDays } from 'date-fns';`

Comment (thomasheartman) [reply src/lib/db/event-store.ts:415]: Ah, yeah, that would work as well! Which ever you think is better 💁🏼  Probably safer to go for 24.

Comment (gastonfournier) [reply src/lib/db/event-store.ts]: Thanks! I added subDays but I'm not 100% confident in my querying skills so I'll write a test

**Review by ivarconr — APPROVED** [review]

**Review by FredrikOseberg — APPROVED** [review]
LGTM

Regarding discussion point: 

The idea is solid, but it will also introduce query complexity. We should investigate whether we can solve the same problem in a different way before we decide on implementation.

**Comment by gastonfournier** [comment]
> Regarding discussion point:
> 
> The idea is solid, but it will also introduce query complexity. We should investigate whether we can solve the same problem in a different way before we decide on implementation.

Another alternative by @gardleopard using partial indexes to index on the announced column (which is boolean but after doing a quick read it doesn't seem to be a problem for partial indexes) https://www.postgresql.org/docs/current/indexes-partial.html

**Comment by gastonfournier** [comment]
We couldn't notice an improvement so we're trying this instead: https://github.com/Unleash/unleash/pull/4856

**Pull request merged**